bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java

```diff
@@ -225,25 +230,25 @@ private List<Path> listFiles(Path path, boolean includeSubDirectory) {
     public void processWatchEvent(WatchService.Kind kind, Path path) {
         Path fullPath = watchPath.resolve(path);
         File file = fullPath.toFile();
-        if (!file.isHidden()) {
-            if (kind == DELETE) {
-                if (file.isDirectory()) {
-                    if (watchSubDirectories) {
-                        synchronized (this) {
-                            String prefix = fullPath.getParent().toString();
-                            Set<String> toRemove = scriptMap.keySet().stream().filter(ref -> ref.startsWith(prefix))
-                                    .collect(Collectors.toSet());
-                            toRemove.forEach(this::removeFile);
-                        }
+
+        if (kind == DELETE) {
+            if (file.isDirectory()) {
+                if (watchSubDirectories) {
+                    synchronized (this) {
+                        String prefix = fullPath.getParent().toString();
+                        Set<String> toRemove = scriptMap.keySet().stream().filter(ref -> ref.startsWith(prefix))
+                                .collect(Collectors.toSet());
+                        toRemove.forEach(this::removeFile);
                     }
-                } else {
-                    removeFile(ScriptFileReference.getScriptIdentifier(file.toPath()));
+                }
+            } else {
+                String scriptIdentifier = ScriptFileReference.getScriptIdentifier(fullPath);
+                if (scriptMap.containsKey(scriptIdentifier)) {
+                    removeFile(scriptIdentifier);
                 }
             }
-
-            if (file.canRead() && (kind == CREATE || kind == MODIFY)) {
-                addFiles(listFiles(file.toPath(), watchSubDirectories));
-            }
+        } else if (!file.isHidden() && file.canRead() && (kind == CREATE || kind == MODIFY)) {
```

Comment:
I moved the hidden file check here, because they should be checked during deletion (above), in case they got added previously. Otherwise they'd never be deleted.